### ISSUE TITLE
could you reopen #697?

### ISSUE THREAD

**Issue by hotbroker** [yhirose/cpp-httplib#699]
after reviewing the code,im pretty sure this is an error:

reusing the socket in SSL_shutdown but the socket is closed before SSL_shutdown  in SSLClient::close_socket

**Comment by yhirose** [comment]
@hotbroker, glad to hear that. Could you show the smallest possible code snippet, so that I can easily reproduce it on my Mac? Thanks!

**Comment by hotbroker** [comment]
> @hotbroker, glad to hear that. Could you show the smallest possible code snippet, so that I can easily reproduce it on my Mac? Thanks!

well ,i dont think you can reproduc this ? 
sometimes this will may not cause a big prolem,like close a invalid handle,or send data with a invalid handle.
but also may cause a serious issue,like close a handle not belong to it(the secon time it was close,this handle value may be created by other thread for othe type resource like event,file etc *on windows*)

**Comment by yhirose** [comment]
@hotbroker, I think the problem should be reproducible regardless the operating system that you use, because it's OpenSSL issue. Could you anyway try to make a smallest possible snippet? Then, I'll debug it. Thanks!

**Comment by hotbroker** [comment]
> @hotbroker, I think the problem should be reproducible regardless the operating system that you use, because it's OpenSSL issue. Could you anyway try to make a smallest possible snippet? Then, I'll debug it. Thanks!

`	std::string url = "https://www.apple.com/";
	Url u(url);
	string ep = u.scheme() + "://" + u.host();

	string path = url.substr(ep.length());

	auto res1 = httplib::Client(ep.data());

	auto res = res1.Get(path.data());
`
check this out.
**stack as follow:**

00 vrfcore!VerifierStopMessageEx
01 vfbasics!AVrfpVectoredExceptionHandler
02 ntdll!RtlpCallVectoredHandlers
03 ntdll!RtlDispatchException
04 ntdll!KiUserExceptionDispatcher
05 ntdll!NtQueryObject
06 KERNELBASE!GetHandleInformation
07 WS2_32!DSOCKET::FindIFSSocket
08 WS2_32!send
09 !sock_write
0a !bwrite_conv
0b !bio_write_intern
0c !BIO_write
0d !ssl3_write_pending
0e !do_ssl3_write
0f !ssl3_dispatch_alert
10 !ssl3_send_alert
11 !ssl3_shutdown
12 !SSL_shutdown
13 !httplib::detail::ssl_delete
14 !httplib::SSLClient::close_socket
15 !httplib::ClientImpl::stop_core
16 !httplib::ClientImpl::process_request
17 !httplib::ClientImpl::handle_request
18 !<lambda_66ee1219a816680a9178e5bd970ecab4>::operator()
19 !std::invoke<<lambda_66ee1219a816680a9178e5bd970ecab4> &,httplib::Stream &>
1a !std::_Invoker_ret<bool,0>::_Call<<lambda_66ee1219a816680a9178e5bd970ecab4> &,httplib::Stream &>
1b !std::_Func_impl_no_alloc<<lambda_66ee1219a816680a9178e5bd970ecab4>,bool,httplib::Stream &>::_Do_call
1c !std::_Func_class<bool,httplib::Stream &>::operator()
1d !httplib::detail::process_client_socket_ssl<std::function<bool __cdecl(httplib::Stream &)> >
1e !httplib::SSLClient::process_socket
1f !httplib::ClientImpl::send
20 !httplib::ClientImpl::Get
21 !httplib::ClientImpl::Get
22 !httplib::Client::Get
23 !test123
24 !CWEMoneyToolDlg::OnInitDialog

**Comment by lightvector** [comment]
@yhirose - it sounds to me like hotbroker is pointing to an instance where a socket is used in a certain way after it is closed - in particular, SSL_shutdown may attempt to send data down the socket in order to carry out the SSL shutdown protocol. 

It could very well be OS-specific whether this actually results in an error being reported or not, because it could be OS-specific what happens when you try to write data into a closed socket (or OS-specific in exactly what errors are returned, and therefore how OpenSSL will react to them).

hotbroker is claiming that a problem occurs on Windows in this case. Even if you can't reproduce an error on Mac, if the fixed code also still works on Mac too, then it seems like just simply a good fix to include? Since SSL_shutdown may attempt to communicate using the socket, unless I'm missing something, it seems like a bug to close the socket before SSL_shutdown, instead of the other way around.

See https://www.openssl.org/docs/man1.0.2/man3/SSL_shutdown.html for a description of the semantics of the function. It sends data over the "bio" (which in this case is the socket) to signal the other party that a shutdown is happening. *After* that, the connection can be closed if desired. 

In the case of wanting to re-use the socket, again as documented, you can call SSL_shutdown a second time in the case that it returns 0, and then SSL will be shutdown but the socket will still be active as just a normal socket. I don't see any reason why httplib would want to do this though, this only applies if for some reason you want to terminate SSL yet still use the socket for non-SSL communication, or something like that, so just SSL_shutdown and then closing the socket should be fine, but should be done in that order (SSL_shutdown, then close socket) instead of the reverse order (close socket, then SSL_shutdown).

**Comment by lightvector** [comment]
Simple fix would presumably be this, similar to what hotbroker posted in the other issue:
```
 inline void SSLClient::close_socket(Socket &socket, bool process_socket_ret) {
-  detail::close_socket(socket.sock);
-  socket_.sock = INVALID_SOCKET;
   if (socket.ssl) {
     detail::ssl_delete(ctx_mutex_, socket.ssl, process_socket_ret);
     socket_.ssl = nullptr;
   }
+  detail::close_socket(socket.sock);
+  socket_.sock = INVALID_SOCKET;
 }
```

**Comment by yhirose** [comment]
@lightvector, thank very much for the detailed research. All you mentioned sound good to me. Could you send your suggested code as a pull request?  Thanks!

**Comment by lightvector** [comment]
It seems that the issue is more complex - fixing this bug in isolation in a proper way is not easy due to the presence of other synchronization issues, so I went ahead and tried to fix the other bugs I found too. I think the pull request I submitted is a correct fix of this and several other issues all together though.

**Comment by hotbroker** [comment]
> It seems that the issue is more complex - fixing this bug in isolation in a proper way is not easy due to the presence of other synchronization issues, so I went ahead and tried to fix the other bugs I found too. I think the pull request I submitted is a correct fix of this and several other issues all together though.

thats great, good to know,i should update my code know .
and the fixing code I posted may not right ,since the socket value is duplicated to diff places,

**Comment by yhirose** [comment]
@hotbroker, I don't fully understand this issue yet, Could please give me another small complete reproducible code? Then, I'll build it on my machine and when I confirm that it occurs on my machine, I'll convert it to a unit test in `test/test.cc` and start debugging it. Thanks for your help!

**Comment by hotbroker** [comment]
> @hotbroker, I still don't fully understand this issue yet. Could please give me another small complete reproducible code? Then, I'll build it on my machine and when I confirm that it occurs on my machine, I'll convert it to a unit test in `test/test.cc` and start debugging it. Thanks for your help!

well,this issue is a little complicated, and lightvector explain on detail.
in summary,when we initial a https ,inside the ssl_delete()->ssl_shutdown socket of  openssl ,sometimes there may have a path that will send some data again with the original socket name socket_ori (refered lightvector  *SSL_shutdown may attempt to send data down the socket in order to carry out the SSL shutdown protocol.*),but as we can see the code I posted ,we have close the socket:socket_ori before ssl_delete

plus,im sorry if something i said confused you,since im not that good at speaking english

**Comment by yhirose** [comment]
Should be fixed by #701.
